pip/pip-385.md

```diff
@@ -0,0 +1,348 @@
+    Server-->>Client: ProducerSuccess(reqId, producerName)
+    Activate Client
+    Activate Server
+    Client->>Server: Send(1, message1)
+    Client->>Server: Send(2, message2)
+    Server-->>Client: SendReceipt(1, msgId1)
+    Client->>Server: Send(3, message3)
+    Client->>Server: Send(4, message4)
+    Note right of Server: Topic breaching quota
+    Activate Server
+    note right of Server: TCP channel read paused
+    Client-xServer: Send(5, message5)
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Client-xServer: Send(6, message6)
+    Server-->>Client: SendReceipt(4,msgId4)
+    note right of Server: TCP channel read resumed
+    deactivate Server
+    Server-->>Server: read message 5
+    Server-->>Server: read message 6
+    Client->>Server: Send(7, message7)
+    Server-->>Client: SendReceipt(5,msgId5)
+    Server-->>Client: SendReceipt(6,msgId6)
+    Server-->>Client: SendReceipt(7,msgId7)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Challenges with the current approach
+
+The current approach may look perfectly fine when looking at the above example, but when looked from a wider scope,
+things start looking bad.
+Typically, the clients reuse a single TCP connection from the client to a broker to send messages to multiple topics.
+This is controlled by the client side property
+of [connectionsPerBroker](https://github.com/apache/pulsar/blob/4b3b273c1c57741f9f9da2118eb4ec5dfeee2220/pulsar-client/src/main/java/org/apache/pulsar/client/impl/conf/ClientConfigurationData.java#L135)
+which defaults to 1. The situation is worsened by the fact that typically, a client is used to create producers for
+partitioned topics and generally an application may produce to more than one partitioned topic with the producers
+created from the same client object, thus all sharing the same tcp connection.
+
+In this situation, even when a single topic starts breaching the quota, the entire TCP connection is paused leading to
+a noisy neighbour effect where effectively all the topics that the client is producing to start getting throttled and
+observe high latencies.
+
+# Motivation
+
+The current method of inducing throttling when a topic or connection breaches quota has various challenges:
+
+* **Noisy neighbors** - Even if one topic is exceeding the quota, since the entire channel read is paused, all topics
+  sharing the same connect (for example - using the same java client object) get rate limited.
+* **Unaware clients** - clients are completely unaware that they are being rate limited. This leads to all send calls
+  taking super long time or simply timing out (assuming shorter send timeouts). If clients were aware, they can either
+  fail fast or induce back-pressure to their upstream.
+* **Impossible debugging** - Since all topics emit the rate limit metric, it is practically impossible to figure out
```

Comment:
we should have server side metrics for topic throttling which should give a clear picture of msgRate and throttling for any further debugging.

Reply:
broker already emit metrics. But it is very difficult to corelate in cases when there is timeout in one topic, but that's happening only because the entire channel read is paused due to quota breach of other topic.
Moreover, in a situation where the pulsar deployment is managed by a central team and there are N number clients.. its not practical that for every client, we start relying on broker metrics and try to manually corelate the noisy neighbours. Even the fix for such cases is, at times, trial and error to see if quota increase of one of the suspected topics helps. This is a very iterative and time taking process.